students/arun_nalla/session04/trigrams.py

```diff
@@ -0,0 +1,50 @@
+﻿#! usr/bin/env python 3
+#words = 'I wish I may I wish I might'.split()
+def trigram_dict (words):
+    trilist = {}
+    for i in range(len(words)-2):
+        a = tuple(words[i:i+2])
+        b = words[i+2]
+        if a not in trilist:
+            trilist[a] = [b]
+        else:
+            trilist[a].append(b)
+    return trilist
+def new_data(old_data):
+    word_list = ''
+    with open(filename, 'r') as file_name:
+        translator = str.maketrans('','', string.punctuation)
+        for lines in file_name.readlines():
+            words = lines.translate(translator)
+            word_list+=words
+    list_of_words = word_list.split()
+    return list_of_words
+import random
+import string
+def build_new_story(story):
+    keys_first = random.choice(list(trigram_dict(words)))
+    #print (keys_first, "key first")
+    my_list = list(keys_first)
+    #print (my_list, "new list")
+    my_list.append(random.choice(trigram_dict(words)[keys_first]))
```

Comment:
You are making `trigram_dict(words)` call many times (here, at line 25, 33, 34, anywhere else?). Note that every time you call `trigram_dict(words)`, Python runtime executes all statements in the function, which is basically reconstructing the same dict again and again. No need for such repeated/duplicate computation.

Call the function once, save the result to a variable, and use the variable whenever it's needed. In other words, put a statement `trigrams = trigram_dict(words)` after line 24, and replace every `trigram_dict(words)` occurrence in this function with `trigrams`.